File: packages/cinira-crypto-java/src/main/kotlin/cinira/util/ASN1Contents.kt
package cinira.util

import java.lang.IllegalStateException

class ASN1Contents private constructor(
    val numbers: List<Int>,
    val oids: List<ByteArray>,
    val strings: List<ByteArray>
) {
    companion object {
        fun fromBytes(bytes: ByteArray): ASN1Contents {
            val numbers = mutableListOf<Int>()
            val oids = mutableListOf<ByteArray>()
            val strings = mutableListOf<ByteArray>()
            parseASN1(bytes, numbers, oids, strings)
            return ASN1Contents(numbers.toList(), oids.toList(), strings.toList())
        }
    }
}

private fun parseASN1(
    buffer: ByteArray,
    numbers: MutableList<Int>,
    oids: MutableList<ByteArray>,
    strings: MutableList<ByteArray>
) {
    var pos = 0
    while (pos < buffer.size) {
        val tag = buffer[pos++].toInt()
        var length = buffer[pos++].toInt()
        if (0 != (length and 0x80)) {
            length = (0 until (length and 0x7f)).fold(0) { extLen, _ ->
                (extLen shl 8) or (buffer[pos++].toInt() and 0xff)
            }
        }
        val contents = buffer.sliceArray(pos until pos + length)
        pos += length
        if (0x30 == tag) {
            parseASN1(contents, numbers, oids, strings)
        } else if (0x03 == tag || 0x04 == tag) {
            strings += contents
        } else if (0x06 == tag) {
            oids += contents
        } else if (0x02 == tag) {
            var value = 0
            (0 until contents.size).forEach { i ->
                value = (value shl 8) or (contents[i].toInt() and 0xff)
            }
            numbers += value
        } else if (0x05 != tag) {
            throw IllegalStateException("Unexpected ASN.1 tag $tag.")
        }
    }
}

File: packages/cinira-crypto-js/src/ASN1Contents.ts
import {freeze} from 'immer';
import _ from "lodash";

/**
 * Parse the raw contents of an ASN.1 buffer into separate arrays if numbers, OIDs, and byte strings. Borrows (very
 * heavily) from https://commandlinefanatic.com/cgi-bin/showarticle.cgi?article=art050
 *
 * @param buffer the ASN.1 buffer to parse.
 * @param numbers the output array to received parsed numbers.
 * @param oids the output array to receive parsed OIDs.
 * @param strings the output array to receive parsed byte strings.
 */
function parseASN1(buffer: Buffer, numbers: number[], oids: Buffer[], strings: Buffer[]) {
    /* tslint:disable:no-bitwise */
    for (let pos = 0; pos < buffer.length;) {
        const tag = buffer[pos++];
        let length = buffer[pos++];
        if (0 !== (length & 0x80)) {
            let extLen = 0;
            for (let i = 0; i < (length & 0x7f); i += 1) {
                extLen = (extLen << 8) | (buffer[pos++] & 0xff);
            }
            length = extLen;
        }
        const contents = buffer.subarray(pos, pos + length);
        pos += length;
        if (0x30 === tag) {
            parseASN1(contents, numbers, oids, strings);
        } else if (0x03 === tag || 0x04 === tag) {
            strings.push(contents);
        } else if (0x06 === tag) {
            oids.push(contents);
        } else if (0x02 === tag) {
            numbers.push(_.reduce(contents, (acc, next) => (acc << 8) | (next & 0xff), 0));
        } else if (0x05 !== tag) {
            throw Error(`Unexpected ASN.1 tag ${tag}.`);
        }
    }
    /* tslint:enable:no-bitwise */
}

/**
 * [ASN1Contents] holds the (naively) parsed contents of an ASN.1 buffer, separated into numbers (ints), OIDs, and byte
 * strings by index within the file.
 */
export class ASN1Contents {
    private constructor(
        public readonly numbers: number[],
        public readonly oids: Buffer[],
        public readonly strings: Buffer[],
    ) {
    }

    /**
     * Parse ASN.1 data from an existing buffer.
     *
     * @param buffer the ASN.1 buffer to parse.
     */
    static fromBuffer(buffer: Buffer) {
        const numbers: number[] = [];
        const oids: Buffer[] = [];
        const strings: Buffer[] = [];
        parseASN1(buffer, numbers, oids, strings);
        return freeze(new ASN1Contents(numbers, oids, strings));
    }
}
